docs/enhancements/bmac-reboot-hosts-on-iso-change.md

```diff
@@ -0,0 +1,131 @@
+* Leave the door open for other platforms (include None/boot-it-yourself) to also avoid host reboots that are not needed.
+
+### Non-Goals
+
+* 
+
+## Proposal
+
+The assisted-service will track whether a particular Agent was booted using an
+old InfraEnv or NMStateConfig. It will ensure that any Agent that was booted
+using obsolete artifacts will have:
+* a label indicating that it was booted from obsolete artifacts
+* an array in its Status of references to the resources that are obsolete
+
+### assisted-service InfraEnv controller
+
+The InfraEnv ISO URL will change when the ISO contents change. This will make
+it obvious to an API user that they have booted a host from an obsolete ISO,
+because the URLs won't match. A side benefit of ensuring that the ISO URL
+uniquely identifies the resource's contents, is that pull-through caches will
+have an easy time understanding when an file is no longer relevant.
+
+When assisted-service creates an ISO, it will capture and include in the ISO:
+* the name, namespace, UID, and generation of the InfraEnv
+* the name, namespace, UID, and generation of each NMStateConfig
+
+At runtime, those resource identifiers for the InfraEnv and the NMStateConfig
+it used will be passed by the agent to assisted-service and compared against
+the current resource generations to determine if they are obsolete.
+
+### assisted-service Agent controller
+
+When the Agent resource gets created, its status should include the
+above-described reference details for:
+* The InfraEnv
+* The NMStateConfig it utilized during its boot
+
+If those details indicate that any of the referenced resources have changed,
+then the controller will add a label to the Agent resource indicating such. It
+will also add a reference to each resource to an array in the Agent's Status.
+
+### Bare Metal Agent Controller
+
+The Bare Metal Agent Controller will watch for Agents to have that label and
+reboot the corresponding BareMetalHost with the latest ISO URL.
+
+The BMAC will also watch for the InfraEnv URL to change, and if there are any
+BareMetalHosts that were booted with an obsolete URL but do not yet have an
+Agent resource, those BareMetalHosts will be rebooted with the latest ISO URL.
+
+### User Stories
+
+#### Story 1
+
+As a user of Zero Touch Provisioning, when I add and modify NMStateConfig
+resources, unaffected hosts will not reboot. Only those hosts directly impacted
+will be rebooted with a current ISO URL.
+
+#### Story 2
+
+As a boot-it-yourself user, I will manually or with automation watch for Agents
+to have the obsolete artifact label and restart them with a fresh ISO.
+
+### Implementation Details/Notes/Constraints [optional]
+
+* Any Agent that has already started provisioning will be exempt from rebooting.
+
+### Risks and Mitigations
+
+
+## Design Details [optional]
+
+
+### Open Questions
+
+* Should the BMAC delete an Agent after it reboots the corresponding host?
```

Comment:
For late binding it must not delete it.
Besides, the agent is a resource of the Assisted-Service, not the BMAC, so I don't think it makes sense for BMAC to delete it

Reply:
I agree.  What's the motivation for having BMAC delete the Agent?

Reply:
We'll stay with the same Agent so no reason to delete it